Improving slicing for nn.ModuleList to support arbitrary step size

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58361 More robust slicing logic for nn.ModuleList**

This PR makes the slicing logic for nn.ModuleList more robust. Previously, it was only supported for step size of 1 . Since we already implemented more robust tuple slicing before where we handle arbitrary step size, we made sure that the slicing logic for nn.ModuleList follows the same pattern as general tuple slicing. 
Fixes: https://github.com/pytorch/pytorch/issues/58360

Differential Revision: [D28464855](https://our.internmc.facebook.com/intern/diff/D28464855)

tugsbayasgalan committed:
Fix moduleList slicing

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28464855).

tugsbayasgalan committed:
Update on "Fix moduleList slicing"


Fixes: https://github.com/pytorch/pytorch/issues/58360

Differential Revision: [D28464855](https://our.internmc.facebook.com/intern/diff/D28464855)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28464855).

tugsbayasgalan committed:
Update on "Fix moduleList slicing"


Fixes: https://github.com/pytorch/pytorch/issues/58360

Differential Revision: [D28464855](https://our.internmc.facebook.com/intern/diff/D28464855)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28464855).

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -4208,20 +4208,29 @@ struct to_ir {
+              NamedValue(val_range, "step", emitExpr(Expr(slice.step().get())));
+          tuple_args.emplace_back(step);
+        } else {
+          tuple_args.emplace_back(c10::nullopt);
```
Is this branch tested?

tugsbayasgalan replied:
done

tugsbayasgalan committed:
Update on "More robust slicing logic for nn.ModuleList"


This PR makes the slicing logic for nn.ModuleList more robust. Previously, it was only supported for step size of 1 . Since we already implemented more robust tuple slicing before where we handle arbitrary step size, we made sure that the slicing logic for nn.ModuleList follows the same pattern as general tuple slicing. 
Fixes: https://github.com/pytorch/pytorch/issues/58360

Differential Revision: [D28464855](https://our.internmc.facebook.com/intern/diff/D28464855)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28464855).

gmagogsfm commented on test/jit/test_slice.py:
```diff
@@ -136,3 +136,25 @@ def tuple_slice(a):
+        scripted_foo = torch.jit.script(Foo())
+        result1, result2 = scripted_foo()
+
+        self.assertEqual(len(result1), 2)
+        self.assertEqual(str(result1[0]._type()), "__torch__.torch.nn.modules.rnn.GRU")
+        self.assertEqual(str(result1[1]._type()), "__torch__.torch.nn.modules.linear.Linear")
+
+        self.assertEqual(len(result2), 1)
+        self.assertEqual(str(result2[0]._type()), "__torch__.torch.nn.modules.activation.ReLU")
```
Type string comparison is fragile, please use a more robust method to verify test results.

One potential solution is to create a dummy module class that has a string-typed attribute. Then you can create 3 instances that each has a different value for that attribute. This is both lighter-weight and more robust than current implementation.

tugsbayasgalan replied:
done

gmagogsfm commented on test/jit/test_slice.py:
```diff
@@ -136,3 +136,25 @@ def tuple_slice(a):
+                self.test = torch.nn.ModuleList([m1, m2, m3])
+
+            def forward(self):
+                return self.test[::-2], self.test[1:2:]
```
These are not great test data.

For example, `[1:2:]` is bad for testing step size, any `step` size would result in same slicing behavior.

```
>>> a = [1, 2, 3]
>>> a[1:2:3]
[2]
>>> a[1:2:2]
[2]
>>> a[1:2:1]
[2]
```

tugsbayasgalan replied:
done

tugsbayasgalan committed:
Update on "More robust slicing logic for nn.ModuleList"


This PR makes the slicing logic for nn.ModuleList more robust. Previously, it was only supported for step size of 1 . Since we already implemented more robust tuple slicing before where we handle arbitrary step size, we made sure that the slicing logic for nn.ModuleList follows the same pattern as general tuple slicing. 
Fixes: https://github.com/pytorch/pytorch/issues/58360

Differential Revision: [D28464855](https://our.internmc.facebook.com/intern/diff/D28464855)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "More robust slicing logic for nn.ModuleList"


This PR makes the slicing logic for nn.ModuleList more robust. Previously, it was only supported for step size of 1 . Since we already implemented more robust tuple slicing before where we handle arbitrary step size, we made sure that the slicing logic for nn.ModuleList follows the same pattern as general tuple slicing. 
Fixes: https://github.com/pytorch/pytorch/issues/58360

Differential Revision: [D28464855](https://our.internmc.facebook.com/intern/diff/D28464855)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28464855).

tugsbayasgalan committed:
Update on "More robust slicing logic for nn.ModuleList"


This PR makes the slicing logic for nn.ModuleList more robust. Previously, it was only supported for step size of 1 . Since we already implemented more robust tuple slicing before where we handle arbitrary step size, we made sure that the slicing logic for nn.ModuleList follows the same pattern as general tuple slicing. 
Fixes: https://github.com/pytorch/pytorch/issues/58360

Differential Revision: [D28464855](https://our.internmc.facebook.com/intern/diff/D28464855)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28464855).
